offload/libomptarget/omptarget.cpp

```diff
@@ -515,6 +675,187 @@ int targetDataBegin(ident_t *Loc, DeviceTy &Device, int32_t ArgNum,
+      bool IsNewlyAllocated =
+          llvm::any_of(AttachInfo.NewAllocations, [&](const auto &Alloc) {
+            void *AllocPtr = Alloc.first;
+            int64_t AllocSize = Alloc.second;
+            return Ptr >= AllocPtr &&
+                   Ptr < (void *)((char *)AllocPtr + AllocSize);
+          });
+      DP("ATTACH entry %zu: %s pointer " DPxMOD " was newly allocated: %s\n",
+         EntryIdx, PtrName, DPxPTR(Ptr), IsNewlyAllocated ? "yes" : "no");
+      return IsNewlyAllocated;
+    };
+
+    // Only process ATTACH if base/begin was newly allocated OR ALWAYS flag is
+    // set
+    if (!IsAttachAlways && !WasNewlyAllocated(HstPtr, "pointer") &&
+        !WasNewlyAllocated(HstPteeBegin, "pointee")) {
+      DP("Skipping ATTACH entry %zu: neither pointer nor pointee was newly "
+         "allocated and no ALWAYS flag\n",
+         EntryIdx);
+      continue;
+    }
+
+    DP("Processing ATTACH entry %zu: Always=%s\n", EntryIdx,
+       IsAttachAlways ? "yes" : "no");
+
+    // Lambda to perform target pointer lookup and validation
+    auto LookupTargetPointer =
+        [&](void *Ptr, int64_t Size,
+            const char *PtrType) -> std::optional<TargetPointerResultTy> {
+      // ATTACH map-type does not change ref-count, or do any allocation
+      // We just need to do a lookup for the pointer/pointee.
+      TargetPointerResultTy TPR = Device.getMappingInfo().getTgtPtrBegin(
+          Ptr, Size, /*UpdateRefCount=*/false,
+          /*UseHoldRefCount=*/false, /*MustContain=*/true);
+
+      DP("ATTACH entry %zu: %s lookup - HstPtr=" DPxMOD ", TgtPtr=" DPxMOD
+         ", IsPresent=%s, IsHostPtr=%s\n",
+         EntryIdx, PtrType, DPxPTR(Ptr), DPxPTR(TPR.TargetPointer),
+         TPR.isPresent() ? "yes" : "no",
+         TPR.Flags.IsHostPointer ? "yes" : "no");
+
+      if (!TPR.isPresent()) {
+        DP("Skipping ATTACH entry %zu: %s not present on device\n", EntryIdx,
+           PtrType);
+        return std::nullopt;
+      }
+      if (TPR.Flags.IsHostPointer) {
+        DP("Skipping ATTACH entry %zu: device version of the %s is a host "
+           "pointer.\n",
+           EntryIdx, PtrType);
+        return std::nullopt;
+      }
+
+      return TPR;
+    };
+
+    // Get device version of the pointer (e.g., &p)
+    auto PtrTPROpt = LookupTargetPointer(HstPtr, PtrSize, "pointer");
+    if (!PtrTPROpt)
+      continue;
+    TargetPointerResultTy &PtrTPR = *PtrTPROpt;
+    void **TgtPtrBase = (void **)PtrTPR.TargetPointer;
+
+    // Get device version of the pointee (e.g., &p[10])
+    auto PteeTPROpt = LookupTargetPointer(HstPteeBegin, 0, "pointee");
+    if (!PteeTPROpt)
+      continue;
+    void *TgtPteeBegin = PteeTPROpt->TargetPointer;
+
+    // Insert a data-fence before the first pointer-attachment.
+    if (IsFirstPointerAttachment) {
+      IsFirstPointerAttachment = false;
+      DP("Inserting a data fence before the first pointer attachment.\n");
+      Ret = Device.dataFence(AsyncInfo);
```

Comment:
This can be delayed until just before the `submitData` call in `performPointerAttachment`, but the code-flow would become a bit uglier and less obvious to read, as the flag would have to be passed by reference into `performPointerAttachment`, and set to `false` in there.